libraries/botbuilder-dialogs-declarative/src/resources/resourceExplorer.ts

```diff
@@ -239,18 +259,15 @@ export class ResourceExplorer {
         return loader.load(value, type) as T;
     }
 
-    /**
-     * @private
-     */
-    private getComponentRegistrations(): ComponentRegistration[] {
-        return ComponentRegistration.components.filter(
-            (component: ComponentRegistration) => 'getDeclarativeTypes' in component
+    private getComponentRegistrations(): ComponentDeclarativeTypes[] {
+        return (
+            this._declarativeTypes ??
+            (ComponentRegistration.components.filter(
+                (component: ComponentRegistration) => 'getDeclarativeTypes' in component
+            ) as ComponentDeclarativeTypes[])
```

Comment:
Why is the `as` casting necessary?

Reply:
The type of `ComponentRegistration.components` is `ComponentRegistration` not `ComponentDeclarativeTypes`, so I have to cast it to `ComponentDeclarativeTypes` to avoid type mismatch.
If a component registration (derived from `ComponentRegistration` class) doesn't implement `ComponentDeclarativeTypes` (doesn't have a _getDeclarativeTypes_ method), it should be filtered out.

Reply:
Is `ComponentDeclarativeTypes` a subclass of `ComponentRegistration`? Or a different type entirely?

Reply:
No, they are different types, checkout corresponding types in botbuilder-dotnet:
[ComponentRegistration](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder/ComponentRegistration.cs)
[IComponentDeclarativeTypes](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Declarative/IComponentDeclarativeTypes.cs)

Reply:
`ComponentRegistration.components` should have a different type signature then, right?

Reply:
`ComponentRegistration.components` should be `ComponentRegistration[]` but each component can implement one or more interfaces like `ComponentDeclarativeTypes` to provide different assets.
Here are some docs in `ComponentRegistration` of C#:
> To make your components available to the system you derive from ComponentRegistration and implement appropriate 
> interfaces which register functionality.  These components then are consumed in appropriate places by the 
> systems that need them. For example, to add declarative types to the system you simply add class that 
> implements IComponentDeclarativeTypes.

Reply:
How are components added to this array? It seems wrong that disjoint types are stored in a single place with only one of the type annotations.

Reply:
They are added via `ComponentRegistration.add`, e.g:
```
ComponentRegistration.add(new AdaptiveComponentRegistration());
ComponentRegistration.add(new LuisComponentRegistration());
```
In C#, there are two other interfaces other than `IComponentDeclarativeTypes`, though they are not required by a `ResourceExplorer`:
* `IComponentMemoryScopes` to provide `MemoryScope`
* `IComponentPathResolvers` to provide `PathResolver`

Reply:
Rather than inlining this type-checking logic, let's extract that to a type guard defined in the same file as the interface (`ComponentDeclarativeType`) in question. 

```typescript
import { tests } from 'botbuilder-stdlib';

export function isComponentDeclarativeType(val: any): val is ComponentDeclarativeType {
  return tests.isObject(val) && tests.isFunc(val.getDeclarativeTypes);
}
```

That should allow you to remove the `as` casting as the type guard informs the Typescript compiler.